stdlib/LinearAlgebra/src/triangular.jl

```diff
@@ -2435,25 +2435,31 @@ function _sqrt_quasitriu_offdiag_block!(R, A)
     return R
 end
 
+# real square root of 2x2 diagonal block of quasi-triangular matrix from real Schur
+# decomposition. Eqs 6.8-6.9 and Algorithm 6.5 of
+# Higham, 2008, "Functions of Matrices: Theory and Computation", SIAM.
 Base.@propagate_inbounds function _sqrt_real_2x2!(R, A)
     a11, a21, a12, a22 = A[1, 1], A[2, 1], A[1, 2], A[2, 2]
+    d = a11 - a22
     θ = (a11 + a22) / 2
-    μ² = -(a11 - a22)^2 / 4 - a21 * a12
-    μ = sqrt(μ²)
-    if θ > 0
-        α = sqrt((sqrt(θ^2 + μ²) + θ) / 2)
-    else
-        α = μ / sqrt(2 * (sqrt(θ^2 + μ²) - θ))
-    end
+    μ = sqrt(-d^2 - 4 * a21 * a12) / 2
+    α = _real_sqrt(θ, μ)
     c = 2α
-    d = α - θ / c
-    R[1, 1] = a11 / c + d
+    f = d / 4α
+    R[1, 1] = α + f
     R[2, 1] = a21 / c
     R[1, 2] = a12 / c
-    R[2, 2] = a22 / c + d
+    R[2, 2] = α - f
     return R
 end
 
+# real part of square root of θ+im*μ
```

Comment:
not sure if it matters here but this function has a sign issue when both parts are negative:

```julia
julia> function _real_sqrt(θ, μ)
           iszero(θ) && iszero(μ) && return sqrt(zero(θ)^2 + zero(μ)^2)
           t = sqrt((abs(θ) + hypot(θ, μ)) / 2)        
           return θ ≥ 0 ? t : μ / 2t
       end
_real_sqrt (generic function with 1 method)

julia> _real_sqrt(-0.3,-0.2)
-0.1740044935431251

julia> real(sqrt(-0.3-0.2im))
0.1740044935431251
```

Reply:
sorry, this is actually not a sign issue, just a choice of which root to choose

Reply:
That's right, yes.